include/rdm/responder.h

```diff
@@ -80,6 +90,92 @@ etcpal_error_t rdmresp_unpack_command(const RdmBuffer *buffer, RdmCommand *cmd);
 bool rdmresp_is_non_disc_command(const RdmBuffer *buffer);
 etcpal_error_t rdmresp_pack_response(const RdmResponse *resp_data, RdmBuffer *buffer);
 
+typedef enum
+{
+  kRespAck = E120_RESPONSE_TYPE_ACK,
+  kRespAckTimer = E120_RESPONSE_TYPE_ACK_TIMER,
+  kRespNackReason = E120_RESPONSE_TYPE_NACK_REASON,
+  kRespAckOverflow = E120_RESPONSE_TYPE_ACK_OVERFLOW,
+  kRespNoSend
+} resp_process_result_t;
+
+typedef enum
+{
+  kRespTypeRDM,
+  kRespTypeController,
+  kRespTypeBroker,
+  kRespTypeDevice
+} resp_type_t;
+
+typedef struct PidHandlerData
+{
+  uint8_t port;
+  uint8_t cmd_class;
+  uint8_t sub_device;
+  size_t overflow_index;
+  void *context;
+
+  const RdmParamData *pd_in;
+  RdmParamData *pd_out;
+} PidHandlerData;
+
+typedef struct RdmPidHandlerEntry
+{
+  uint16_t pid;
+  resp_process_result_t (*handler)(PidHandlerData *data);
+  uint32_t flags;
+} RdmPidHandlerEntry;
+
+typedef struct GetNextQueueMessageData
+{
+  uint8_t port;
+  uint8_t status_type;
+  uint8_t *response_type;
+  uint16_t *sub_device;
+  uint8_t *cmd_class;
+  uint16_t *param_id;
+  RdmParamData *pd;
+} GetNextQueueMessageData;
+
+typedef struct RdmResponderState
+{
+  uint8_t port_number;
+  RdmUid uid;
+  uint8_t number_of_subdevices;
+  resp_type_t responder_type;
+  void *callback_context;
+
+  const RdmPidHandlerEntry *handler_array;
+  size_t handler_array_size;
+
+  uint8_t (*get_message_count)();
+  void (*get_next_queue_message)(GetNextQueueMessageData *data);
+
+} RdmResponderState;
+
+void rdmresp_sort_handler_array(RdmPidHandlerEntry *handler_array, size_t handler_array_size);
+bool rdmresp_validate_state(const RdmResponderState *state);
+
+typedef enum
+{
+  kRespValid,
+  kRespInvalidRdm,
+  kRespNotRdm
+} resp_valid_result_t;
+resp_valid_result_t rdmresp_validate_packet(RdmBufferConstRef buffer, uint8_t calc_checksum);
+
+resp_process_result_t rdmresp_process_packet(const RdmResponderState *state, RdmBufferConstRef bufferIn,
+                                             RdmBufferRef bufferOut, bool *no_break);
```

Comment:
I am not sure passing these structs by value will have the effect you're looking for, doesn't the calling code need to know the length of the data in the output buffer somehow?

Reply:
Good catch, my mistake. Will fix.

Reply:
Done.